crates/flipperzero/src/notification/notes.rs

```diff
@@ -0,0 +1,120 @@
+use super::NotificationMessage;
+
+pub static CLICK: NotificationMessage = NotificationMessage::sound_on(1.0, 1.0);
+
+pub const C0: NotificationMessage = NotificationMessage::sound_on(16.35, 1.0);
```

Comment:
These all hard-code a volume of `1.0`, and AFAICT there is no way to alter this without calling `NotificationMessage::sound_on` yourself with the same frequency constant (which is opaque).

Instead, we could define these constants using the following API:
```rust
/// A pitch at which Flipper Zero can play a sound.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Pitch {
    frequency: f32,
}

impl Pitch {
    /// Defines a new pitch with the given frequency in Hertz.
    /// 
    /// # Safety
    /// 
    /// `frequency` must be positive and greater than zero.
    pub const fn new(frequency: f32) -> Self {
        Self { frequency }
    }

    /// Generates a notification message to start playing sound at this pitch and the
    /// given volume.
    ///
    /// The sound will continue until either another sound replaces it, or [`SOUND_OFF`]
    /// is used.
    /// 
    /// # Safety
    /// 
    /// `volume` must be between `0.0` and `1.0`.
    pub const fn play(self, volume: f32) -> NotificationMessage {
        NotificationMessage::sound_on(self.frequency, volume)
    }
}

pub const C0: Pitch = Pitch::new(16.35);
```

and then use it like:
```rust
pub const SUCCESS: NotificationSequence = notification_sequence![
    messages::DISPLAY_BACKLIGHT_ON,
    messages::GREEN_255,
    messages::VIBRO_ON,
    notes::C5.play(1.0),
    messages::DELAY_50,
    messages::VIBRO_OFF,
    notes::E5.play(1.0),
    messages::DELAY_50,
    notes::G5.play(1.0),
    messages::DELAY_50,
    notes::C6.play(1.0),
    messages::DELAY_50,
    messages::SOUND_OFF,
];
```

---

I also experimented locally with an additional API:
```rust
/// A sound that Flipper Zero can play.
#[derive(Clone, Copy, Debug, PartialEq)]
pub struct Sound {
    pitch: Pitch,
    duration: u32,
}

impl Sound {
    /// Defines a new sound with the given pitch, and duration (in milliseconds).
    pub const fn new(pitch: Pitch, duration: u32) -> Self {
        Self { pitch, duration }
    }

    /// Generates a sequence of notification messages that will play this sound once at
    /// the given volume.
    /// 
    /// # Panics
    /// 
    /// Panics if `volume < 0.0` or `volume > 1.0`.
    pub const fn play_once(self, volume: f32) -> SingleSound {
        (
            self.pitch.play(volume),
            NotificationMessage::delay(self.duration),
            NotificationMessage::sound_off(),
        )
    }
}

// Only defined to make the tuple more ergonomic.
pub type SingleSound = (NotificationMessage, NotificationMessage, NotificationMessage);

const C5_50: SingleSound = Sound::new(notes::C5, 50).play_once(1.0);
const E5_50: SingleSound = Sound::new(notes::E5, 50).play_once(1.0);
const G5_50: SingleSound = Sound::new(notes::G5, 50).play_once(1.0);
const C6_50: SingleSound = Sound::new(notes::C6, 50).play_once(1.0);
pub const SUCCESS: NotificationSequence = notification_sequence![
    messages::DISPLAY_BACKLIGHT_ON,
    messages::GREEN_255,
    messages::VIBRO_ON,
    C5_50.0, C5_50.1,
    messages::VIBRO_OFF,
    E5_50.0, E5_50.1,
    G5_50.0, G5_50.1,
    C6_50.0, C6_50.1,
    messages::SOUND_OFF,
];
```

It should be possible to rework the `notification_sequence!` macro to make this ergonomic:
```rust

impl Sound {
    /// Generates a sequence of notification messages that will play this sound for at
    /// least its specified duration at the given volume.
    ///
    /// The sound will continue after its specified duration until either another sound
    /// replaces it, or [`SOUND_OFF`] is used.
    /// 
    /// # Panics
    /// 
    /// Panics if `volume < 0.0` or `volume > 1.0`.
    pub const fn play(self, volume: f32) -> (NotificationMessage, NotificationMessage) {
        (
            self.pitch.play(volume),
            NotificationMessage::delay(self.duration),
        )
    }
}

/// TODO: Change notification_sequence! to support flattening nested tuples or something.

const C5_50: Sound = Sound::new(notes::C5, 50);
const E5_50: Sound = Sound::new(notes::E5, 50);
const G5_50: Sound = Sound::new(notes::G5, 50);
const C6_50: Sound = Sound::new(notes::C6, 50);
pub const SUCCESS: NotificationSequence = notification_sequence![
    messages::DISPLAY_BACKLIGHT_ON,
    messages::GREEN_255,
    messages::VIBRO_ON,
    play_sound!(C5_50, 1.0),
    messages::VIBRO_OFF,
    play_sound!(E5_50, 1.0),
    play_sound!(G5_50, 1.0),
    play_sound!(C6_50, 1.0),
    messages::SOUND_OFF,
];
```

Reply:
This is a great idea, I was thinking of something similar myself as I have written a [toy music crate](https://github.com/Strosel/rs-music) previously. IMHO the last example seems the best, but I would prefer if the duration and volume were swapped as the volume feels more intrinsic to the note than its duration. Another option would be to give the `play_sound` macro optional volume and duration arguments with default values.

Reply:
After playing with this for a while I seem to be stuck on the `play_sound` macro as it can't "splat" the tuple of messages in a way that makes them usable for `notification_sequence!` any ideas?

What might be best for now is to keep the explicit API of "sound, delay, new sound/sound off" but add a convenience function for generating `sound_on` messages with the desired pitch and volume

Reply:
> I would prefer if the duration and volume were swapped as the volume feels more intrinsic to the note than its duration

I had the reverse feeling, but not strongly; the thing I mainly wanted was the existence of `Pitch` to allow reusing them. Given that `NotificationMessage` groups them as `((frequency, volume), duration)` it also likely makes the macros easier to group them that way.

> After playing with this for a while I seem to be stuck on the `play_sound` macro as it can't "splat" the tuple of messages in a way that makes them usable for `notification_sequence!` any ideas?

Yeah, I encountered the same issue in my brief period of playing with the API (and so I left it as experimental). I think it might be possible by making `play_sound` not be a separate macro, but instead a keyword inside the `notification_sequence!` macro (and then using recursion to parse the macro's arguments left-to-right, and splat the `play_sound` tuple when we reach it).

> What might be best for now is to keep the explicit API of "sound, delay, new sound/sound off" but add a convenience function for generating `sound_on` messages with the desired pitch and volume

Sounds good 👍

Reply:
More disappointing news 😕 due to rust-lang/rust#57241 we can't calculate the frequency from some theoretical `Pitch` enum so I think even a convenience function is impossible at this time. It could be nice to come back to when the issue is resolved though

Reply:
Hehe, I was [already in that issue](https://github.com/rust-lang/rust/issues/57241#issuecomment-1432469599) as a result of this PR, asking if we can have a little `const PartialEq`, as a treat.

Do we need to calculate frequencies though? I think the existing constant-based approach is fine, just using `Pitch` constants instead of `NotificationMessage` constants. If we also want to provide constants for pitch+volume like the C API does, that could be done. But even with an enum approach for a canonical set of sorted notes, the frequencies for them could just be constant.

Reply:
Could you please clarify how we are to use `Pitch` constants instead of `NotificationMessages` when we can not convert one into the other due to restrictions on floats in const?

Reply:
`Pitch::play(self, volume)` is precisely this conversion. I don't understand what float calculations you are seeing as necessary?

Reply:
I think there may have been a misunderstanding on my part. I thought the examples you gave above were more separate and interpreted `Pitch` in the later ones as something like
```rust
enum Pitch {
    C,
    Cs,
    D,
    ...
}
```
which would need float calculations to generate a frequency.

While implementing `Pitch` as only a frequency wrapper like above would be possible (and nice) I think the name could be changed to prevent this confusion for other users. Maybe something like this could be implemented instead

```rust
#[derive(Debug, Clone, Copy)]
pub struct Sound {
    frequency: f32,
    volume: f32,
    duration: u32,
}

impl Sound {
    pub const fn new(frequency: f32) -> Self {
        Self {
            frequency,
            volume: 1.0,
            duration: 0,
        }
    }

    pub const fn volume(mut self, volume: f32) -> Self {
        self.volume = volume;
        self
    }

    pub const fn duration(mut self, duration: u32) -> Self {
        self.duration = duration;
        self
    }

    pub const fn play_once(self) -> NotificationMessage {
        NotificationMessage::sound_on(self.frequency, self.volume)
    }

    pub const fn play(self) -> [NotificationMessage; 3] {
        [
            self.play_once(),
            NotificationMessage::delay(self.duration),
            NotificationMessage::sound_off(),
        ]
    }
}
```
then a constructor using a `Pitch` enum like above could be implemented in the future and the more challenging part right now would "only" be splatting